src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs

```diff
@@ -23,5 +25,13 @@ internal class AcquireTokenCommonParameters
         public IAuthenticationScheme AuthenticationScheme { get; set; } = new BearerAuthenticationScheme();
         public IDictionary<string, string> ExtraHttpHeaders { get; set; }
         public PoPAuthenticationConfiguration PopAuthenticationConfiguration { get; set; }
+
+        /// <summary>
+        /// If set, the client assertion parameters should be ignored and these parameters should be set.
+        /// The input string is the token endpoint and a cancellation token.
+        /// The output is a list of name / value strings to be sent to the HTTP request to the token endpoint
+        /// </summary>
+        public Func<string, CancellationToken, Task<IReadOnlyList<KeyValuePair<string, string>>>> ClientAssertionOverrideDelegate { get; set; }
```

Comment:
I agree. I will make a change. I was pondering between: 

1. A delegate as you mention
2. An interface

# The interface 

```csharp
public interface IClientAssertionProvider
{
    Task<IReadOnlyList<KeyValuePair<string, string>>> GetAssertionAsync(string clientId, string tokenEndpoint, CancellationToken ct);
}
```


```csharp
public Task<IReadOnlyList<KeyValuePair<string, string>>> GetClientAssertionParametersAsync(string clientId, string tokenEndpoint, CancellationToken cancellationToken)
    {
        Assert.AreEqual("https://login.microsoftonline.com/tid/oauth2/v2.0/token", tokenEndpoint);
        IReadOnlyList<KeyValuePair<string, string>> result = new List<KeyValuePair<string, string>>()
                {
                    new KeyValuePair<string, string>("client_assertion", CreateAssertion(clientId, tokenEndpoint)),
                    new KeyValuePair<string, string>("client_assertion_type", "jwt-beaer"),
                };
        return Task.FromResult(result);
    }
```

```csharp
public string GetToken()
{

    var assertionProvider = new MyAssertionProvider();
    var result = await cca.AcquireTokenForClient(scope).WithClientAssertion(assertionProvider).ExecuteAsync();
    return result;
}
```

# Delegate

```csharp
public string GetToken()
{
    var assertionProvider = new MyAssertionProvider();
    var result = await cca.AcquireTokenForClient(scope).WithClientAssertion(GetClientAssertionParametersAsync).ExecuteAsync();
    return result;
}

private static Task<IReadOnlyList<KeyValuePair<string, string>>> GetClientAssertionParametersAsync(string clientId, string tokenEndpoint, CancellationToken cancellationToken)
    {
        Assert.AreEqual("https://login.microsoftonline.com/tid/oauth2/v2.0/token", tokenEndpoint);
        IReadOnlyList<KeyValuePair<string, string>> result = new List<KeyValuePair<string, string>>()
                {
                    new KeyValuePair<string, string>("client_assertion", CreateAssertion(clientId, tokenEndpoint)),
                    new KeyValuePair<string, string>("client_assertion_type", "jwt-beaer"),
                };
        return Task.FromResult(result);
    }

```

Reply:
I think I like the delegate way better, because if we need to support a different delegate the code will be simpler, whearas the interface we will have to create a new one.

Reply:
And is `IReadOnlyList<KeyValuePair<string, string>` needed because we care about the order of the parameters?
Or would a `IReadonlyDictionary<string, string>` work as well?

Reply:
We do not care about the order indeed. I have already moved to using IReadonlyDictionary.